presto-main/src/main/java/com/facebook/presto/server/ConnectorMetadataUpdateHandleJsonSerde.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server;
+
+import com.facebook.airlift.json.JsonCodec;
+import com.facebook.presto.spi.ConnectorMetadataUpdateHandle;
+import com.facebook.presto.spi.connector.ConnectorMetadataUpdateHandleSerde;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import io.airlift.slice.Slices;
+
+import static com.facebook.airlift.json.JsonCodec.jsonCodec;
+
+/**
+ * Json based connector handle serde
+ */
+public class ConnectorMetadataUpdateHandleJsonSerde
+        implements ConnectorMetadataUpdateHandleSerde
+{
+    private LoadingCache<Class, JsonCodec> jsonCodecCache = CacheBuilder.newBuilder()
+            .recordStats()
+            .maximumSize(10_000)
+            .build(CacheLoader.from(cacheKey -> jsonCodec(cacheKey)));
+
+    @Override
+    public byte[] serialize(ConnectorMetadataUpdateHandle val)
+    {
+        JsonCodec jsonCodec = jsonCodecCache.getUnchecked(val.getClass());
```

Comment:
We probably need concrete types for `JsonCodec<XXX>`

Reply:
The tricky bit here is that the exact type is not known during build time and types are not reified.
If I were to refactor and keep this as 
`JsonCodec<? extends ConnectorMetadataUpdateHandle> jsonCodec = jsonCodecCache.getUnchecked(value.getClass());`
the `toBytes` call in the next line would complain since it expects a concrete type.
Let me know if you know of a better way around this.